### PR TITLE
fix: display error message correctly

### LINKED ISSUE #3309: The expected and received values ​​seem to be reversed
### Describe the bug

```expect(xx).toBe(xx)``` 
When an error occurs, the expected value and received value prompt information seem to be reversed.

vitest

image

jest

image

### Reproduction

https://stackblitz.com/edit/node-w1fgff?file=package.json

### System Info

```shell
System:
    OS: Windows 10 10.0.19044
    CPU: (8) x64 Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
    Memory: 2.54 GB / 15.86 GB
  Binaries:
    Node: 16.17.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 7.22.0 - ~\AppData\Roaming\npm\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (112.0.1722.68)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    @vitest/coverage-c8: ^0.29.8 => 0.29.8
    vite: ^4.2.1 => 4.2.1
    vitest: ^0.31.0 => 0.31.0
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by btea**
fix #3309

### CODE REVIEW

File: packages/utils/src/descriptors.ts [hunk 1]
```diff
@@ -94,5 +94,5 @@ export function getConcordanceTheme() {
 }
 
 export function diffDescriptors(actual: unknown, expected: unknown, options: DisplayOptions): string {
-  return concordanceModule.diff(expected, actual, options)
+  return concordanceModule.diff(actual, expected, options)
```
Comment (btea) [inline packages/utils/src/descriptors.ts:97]: Refer to the following code, the first parameter should be actual, the second parameter should be expected

https://github.com/concordancejs/concordance/blob/791d2a89b40eb13f2c889ac270dd8be190cf8073/lib/diff.js#L388-L389

However, the test failed after the modification. According to the error message, it seems to be related to the content of `html.meta.json.gz`. I don't know how to deal with this?

Comment (sheremet-va) [reply packages/utils/src/descriptors.ts:97]: Run vitest with `-u` flag and update snapshots then

Comment (btea) [reply packages/utils/src/descriptors.ts:97]: Some files seem to be deleted after running, I'm not sure if this is correct.

Comment (sheremet-va) [reply packages/utils/src/descriptors.ts:97]: You should've run it only inside the test package that was failing, not all test packages,

Comment (btea) [reply packages/utils/src/descriptors.ts:97]: According to the error message, it seems that the two tests `mocked.test.ts` and `diff.test.ts` failed, but I ran these two tests locally and passed. I don’t know if there is a problem with my local environment. 😕

Comment (sheremet-va) [reply packages/utils/src/descriptors.ts:97]: Seems like you have some problems with your local env, check your branch, and rebuild the project.

Comment (btea) [reply packages/utils/src/descriptors.ts:97]: It seems to be ok after reinstalling dependencies and rebuilding.

File: test/core/test/__snapshots__/mocked.test.ts.snap [hunk 2]
```diff
@@ -45,22 +45,22 @@ Received:
     1st spy call return:
 
     {
-  -   a: '4',
-  +   a: '1',
+  -   a: '1',
```
Comment (sheremet-va) [inline test/core/test/__snapshots__/mocked.test.ts.snap:48]: This is now incorrect. "1" should be received

Comment (btea) [reply test/core/test/__snapshots__/mocked.test.ts.snap:48]: `expect(a).toBe(b)` a is expected, b is received?

Comment (sheremet-va) [reply test/core/test/__snapshots__/mocked.test.ts.snap:48]: It's not expected to be, it's what was called (expected) compared to what was received (actual)

The function was called with 1, but diff shows it was called with 4

Comment (btea) [reply test/core/test/__snapshots__/mocked.test.ts.snap:48]: After this PR change it seems to become called(received) compared to what was expected, so this is correct.

Comment (sheremet-va) [reply test/core/test/__snapshots__/mocked.test.ts.snap:48]: I don't know what you mean, but the diff here is incorrect. The function returned `{ a: '1' }`, but test expected `{ a: '4' }`. So the correct diff is:

```
- a: '4'
+ a: '1'
```

Comment (btea) [reply test/core/test/__snapshots__/mocked.test.ts.snap:48]: I'm a little dizzy now 😵‍💫

According to your description, the tests seem to be reversed.

Comment (sheremet-va) [reply test/core/test/__snapshots__/mocked.test.ts.snap:48]: With your change it is reversed, yes, but it shouldn't be.

Comment (btea) [reply test/core/test/__snapshots__/mocked.test.ts.snap:48]: Sorry, I'm not sure what the correct modification is now. Maybe just reverse the colors here?
https://github.com/vitest-dev/vitest/blob/9836ccb48c5a39c0999484f55731f53ba494be17/packages/utils/src/descriptors.ts#L26-L27

Comment (sheremet-va) [reply test/core/test/__snapshots__/mocked.test.ts.snap:48]: > Sorry, I'm not sure what the correct modification is now. Maybe just reverse the colors here?

Ok, I see now. Give me a bit more time to think

**Comment by sheremet-va** [comment]
You are correct, colors should be changed here:

https://github.com/vitest-dev/vitest/blob/9836ccb48c5a39c0999484f55731f53ba494be17/packages/utils/src/descriptors.ts#L26-L27

**Comment by btea** [comment]
> You are correct, colors should be changed here:
> 
> https://github.com/vitest-dev/vitest/blob/9836ccb48c5a39c0999484f55731f53ba494be17/packages/utils/src/descriptors.ts#L26-L27
> 

Keep the current modified state, and then update the color?

> The spy diff should also be updated with a prefix (but it's probably out of scope):
> 
> ```
> expect(red{vi.fn()}).toName(green{expected})
> ```

Do all tests in the `mocked.test.ts`  file have to add this prefix?

**Comment by sheremet-va** [comment]
> Keep the current modified state, and then update the color?

Yes

> Do all tests in the mocked.test.ts file have to add this prefix?

Just forget about this for now

**Comment by ghiscoding** [comment]
Posting this just an info, this PR seems to be the inverse of recent PR #3289, is that the intent of this PR?

Also looking at possible improvement on the colors, if we run the same test in Jest and Vitest. Jest is coloring in green all the expectations (not just the value but also the test title in green), but in Vitest that entire portion (not just the title) is yellow. I prefer Jest UI a bit more, I find it easier to read and spot the expectation error. Also this PR doesn't correct the red/green missing coloring on the expected value found, does it?

image

in Vitest, both result values are yellow, I thought this PR would fix this?
image

**Comment by btea** [comment]
> Posting this just an info, this PR seems to be the inverse of recent PR #3289, is that the intent of this PR?

I seem to have made a wrong modification, maybe I should just update the colors. This is the result after changing

image

> in Vitest, both result values are yellow, I thought this PR would fix this? Actually this is more complicate than I taught, the text coloring in the expectation comes from theme `getConcordanceTheme()` and is different per variable type. It seems a bit complicate to change

Yes, this PR does not deal with the color problem of the value, it can be improved later.

**Comment by sheremet-va** [comment]
> This is the result after changing

The string diff is confusing now (the red bg is in green "minus" sign) 😞 

We should probably improve the diff overall. The concordance tools don't seem to meet our expectations, but jest-diff doesn't print good string diff (line by line, not character by character), which also doesn't meet our expectations

**Comment by ghiscoding** [comment]
@btea I think he's talking about string diffs, a removed character is currently showing with red background and white text, while an addition is showing as green background and white text. So I assume that his comment meant that you somehow changed these colors? Take this for example with current code, how does it show with your new code?

image

**Comment by btea** [comment]
@ghiscoding Oh, thanks for the detailed explanation. I think I now understand what he means, and now there is indeed a problem as he said. Here is how the new code behaves in this example.



Maybe the correct string diff should look like this?



I just tried to modify the color of `delete` and `insert`, not sure if there are other problems. 
https://github.com/vitest-dev/vitest/blob/0f0d75631c3b9d6f335d822a24e25b46b2219eda/packages/utils/src/descriptors.ts#L69-L76

**Comment by ghiscoding** [comment]
@btea which chars are additions and which ones are removal? I don't think the descriptor colors should be touched, because if you scroll up you'll see this comment (no adjustment were made)

https://github.com/vitest-dev/vitest/blob/0f0d75631c3b9d6f335d822a24e25b46b2219eda/packages/utils/src/descriptors.ts#L13-L18

**Comment by btea** [comment]
> @btea which chars are additions and which ones are removal? 

To be honest, I'm not sure what the correct modification is now. Maybe I was wrong in the first place. 😵
>I don't think the descriptor colors should be touched, because if you scroll up you'll see this comment (no adjustment were made)

You are right, I just noticed this comment.

**Pull request merged**